src/core/utils/useGetReservationGroups.ts

```diff
@@ -0,0 +1,141 @@
+ * individual records into a single object.
+ *
+ * We still need the individual reservations and their related records for
+ * further handling so we store these in a new map.
+ *
+ * All other values are for intent and purposes expected to the same for all
+ * reservations in the group.
+ */
+export type ReservationGroupDetails = Omit<
+  ReservationDetailsV2,
+  "reservationId" | "recordId"
+> & {
+  records: {
+    // Record id to reservation id map.
+    [key: string]: number;
+  };
+};
+
+function groupReservations(data: ReservationDetailsV2[]) {
+  const reservationGroups = groupBy(
+    data,
+    (reservation) => reservation.transactionId
+  );
+
+  const processedReserations = map(
+    reservationGroups,
+    (reservationGroup: ReservationDetailsV2[]): ReservationGroupDetails => {
+      return {
+        // Use the first reservation in the group as the base.
+        ...reservationGroup[0],
+        // The queue number for the group is the lowest individual queue number
+        numberInQueue: min(map(reservationGroup, "numberInQueue")),
+        records: reduce(
+          reservationGroup,
+          (result, reservation) => {
+            return {
+              ...result,
+              [reservation.recordId]: reservation.reservationId
+            };
+          },
+          {}
+        )
+      };
+    }
+  );
+  return processedReserations;
+}
+
+/**
+ * Custom version of the generated useGetReservations hook which returns our
+ * custom reservation details type.
+ */
+type UseGetReservationGroupsResult = Omit<
+  ReturnType<typeof useGetReservationsV2>,
+  "data"
+> & {
+  data: ReservationGroupDetails[] | null;
+};
+const useGetReservationGroups = (): UseGetReservationGroupsResult => {
+  const result = useGetReservationsV2();
+  const resultWithGroups = {
+    ...result,
+    data: result.data ? groupReservations(result.data) : null
+  };
+  return resultWithGroups;
+};
+
+export default useGetReservationGroups;
+
+if (import.meta.vitest) {
+  const { describe, expect, it } = import.meta.vitest;
+
+  describe("Reservation grouping", () => {
+    // Generator for setting up mock data with sensible defaults.
+    function generateReservation(
+      data: Partial<ReservationDetailsV2>
+    ): ReservationDetailsV2 {
+      return {
+        reservationId: 123,
+        transactionId: "transaction",
+        recordId: "record",
+        reservationType: "normal", // "normal" or "parallel"
+        state: "reserved", // "reserved", "readyForPickup"
+        dateOfReservation: "2023-01-01",
+        pickupBranch: "DK-775100",
+        expiryDate: "2024-01-01",
+        ...data
+      };
+    }
+
+    it("groups reservations by transaction id", () => {
+      const reservations = groupReservations([
+        generateReservation({ transactionId: "t1" }),
+        generateReservation({ transactionId: "t2" }),
+        generateReservation({ transactionId: "t2" }),
+        generateReservation({ transactionId: "t3" })
```

Comment:
This is an example of what I think we could achieve with something like fishery reload/dpl-react#47 - a generator/factory which ensures we have valid business objects and test cases which can focus on the attributes which are relevant to the case at hand.